### PR TITLE
Case 19807: Switch to vcpkg for many dependencies on all non-mobile platforms

### PR DESCRIPTION
**Pull request by jherico**
## Description

This PR modifies our mechanism for external dependencies.  My current release project will involve adding at least 3 new external dependencies to the Interface build.  However, our current mechanism for external dependencies has a number of issues that this will only exacerbate.

Currently we checkout and build "inline" many upstream projects.  For continuous integration, these significantly add to the build time, because the build directory is wiped for every build.  For developers who maintain multiple checkouts and multiple build directories these significantly add to the total disk space consumed by builds, since each set of external dependencies can take many gigabytes.

The new system uses an open source tool [vcpkg](https://github.com/Microsoft/vcpkg).  The vcpkg tool actually works very similarly to our own custom external build mechanism.  In the vcpkg [ports](https://github.com/Microsoft/vcpkg/tree/master/ports) folder are a large number of projects, and each one has a `CONTROL` file with a name, version and description, as well as a `portfile.cmake` with instructions on how to fetch, build and install the package.  The primary repository contains many packages we don't care about, is missing some packages we do care about, and has versions of some packages that are different than what we require.  Fortunately, the recommended solution to this is to simply [fork](https://github.com/highfidelity/vcpkg) the repository and customize it.  

In order to simplify our process even further, I've written a `prebuild.py` script, which automatically checks out a specific vcpkg commit, bootstraps it, and then builds our required dependencies.  All of this is done either in a temp directory, or the directory pointed to by the `HIFI_VCPKG_BASE` environment variable.  If these actions have already been taken on a given machine for a given vcpkg commit, then the script takes only a couple of second to run.

This approach will solve both the issues with CI as well as for our developers.  In continuous integration, the first time a given vcpkg commit ID is encountered, it will add perhaps a half hour to the build time, but all subsequent builds on that host that use the same set of dependencies will actually have much shorter build times since the dependencies will already be present.  For developers, multiple checkouts and multiple build directories per checkout can all share the same external dependencies.  Additionally,  while the vcpkg build directories can be very large, once built, it's possible to clean up all the build artifacts and reduce the disk footprint from many gigabytes to a few hundred megabytes.

This PR migrates the following external dependencies:

* bullet
* draco
* etc2comp
* gli
* glm
* glslang (new)
* json
* nvtt
* OpenSSL
* OpenVR
* sdl2
* spirv-tools (new)
* spirv-cross (new)
* tbb
* zlib

manuscript sub step: https://highfidelity.fogbugz.com/f/cases/19807/INtroducing-the-vcpkg-build-system

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14681/HighFidelity-Beta-PR14297-b91cc36.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14681/HighFidelity-Beta-Interface-PR14297-b91cc36.exe

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14683/HighFidelity-Beta-PR14297-0dda65d.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14683/HighFidelity-Beta-Interface-PR14297-0dda65d.exe

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14685/HighFidelity-Beta-PR14297-4d99c22.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14685/HighFidelity-Beta-Interface-PR14297-4d99c22.exe

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14698/HighFidelity-Beta-PR14297-7c7e5f6.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14698/HighFidelity-Beta-Interface-PR14297-7c7e5f6.exe

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14734/HighFidelity-Beta-PR14297-8c68835.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14734/HighFidelity-Beta-Interface-PR14297-8c68835.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/14737/HighFidelity-Beta-PR14297-e5023f3.apk

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/14751/HighFidelity-Beta-PR14297-259e6ac.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14751/HighFidelity-Beta-PR14297-259e6ac.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14751/HighFidelity-Beta-Interface-PR14297-259e6ac.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/14772/HighFidelity-Beta-PR14297-ba2b70b.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14772/HighFidelity-Beta-PR14297-ba2b70b.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14772/HighFidelity-Beta-Interface-PR14297-ba2b70b.exe

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14777/HighFidelity-Beta-PR14297-6143a39.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14777/HighFidelity-Beta-Interface-PR14297-6143a39.exe

**Comment by jherico** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14820/HighFidelity-Beta-PR14297-dbd2796.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14820/HighFidelity-Beta-Interface-PR14297-dbd2796.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14825/HighFidelity-Beta-PR14297-8cb0bbe.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14825/HighFidelity-Beta-Interface-PR14297-8cb0bbe.dmg

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/14831/HighFidelity-Beta-PR14297-d9761df.apk

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/14835/HighFidelity-Beta-PR14297-ec835b4.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14831/HighFidelity-Beta-PR14297-d9761df.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14831/HighFidelity-Beta-Interface-PR14297-d9761df.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14835/HighFidelity-Beta-PR14297-ec835b4.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14835/HighFidelity-Beta-Interface-PR14297-ec835b4.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14835/HighFidelity-Beta-PR14297-ec835b4.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14835/HighFidelity-Beta-Interface-PR14297-ec835b4.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/14847/HighFidelity-Beta-PR14297-07f932a.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14847/HighFidelity-Beta-PR14297-07f932a.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14847/HighFidelity-Beta-Interface-PR14297-07f932a.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14847/HighFidelity-Beta-PR14297-07f932a.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14847/HighFidelity-Beta-Interface-PR14297-07f932a.exe

**Comment by qaAdamSmith82** [comment]
Per discussion with dev, this PR requires more than happy path testing

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/14873/HighFidelity-Beta-PR14297-fa000f1.apk

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/14883/HighFidelity-Beta-PR14297-66feac1.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14883/HighFidelity-Beta-PR14297-66feac1.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14883/HighFidelity-Beta-Interface-PR14297-66feac1.dmg

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/14901/HighFidelity-Beta-PR14297-c488b09.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14901/HighFidelity-Beta-PR14297-c488b09.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14901/HighFidelity-Beta-Interface-PR14297-c488b09.exe

File: BUILD.md [hunk 1]
```diff
@@ -82,9 +80,22 @@ In the examples below the variable $NAME would be replaced by the name of the de
...
 ### Optional Components
 
+#### Build Options 
+
+The following build options can be used when running CMake
+
+* BUILD_CLIENT
+* BUILD_SERVER
+* BUILD_TESTS
+* BUILD_TOOLS
```
Comment (AndrewMeadows) [inline BUILD.md]: It looks like you also added `BUILD_MANUAL_TESTS` but failed to mention it here.

File: prebuild.py [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+            f.write(self.id)
+
+        configFilePath = os.path.join(args.build_root, 'vcpkg.cmake')
+        print("Writing cmake config to {}".format(configFilePath))
+        # Write out the configuration for use by CMake
+        cmakeScript = os.path.join(self.path, 'scripts/buildsystems/vcpkg.cmake')
+        installPath = os.path.join(self.path, 'installed', self.triplet)
+        toolsPath = os.path.join(self.path, 'installed', self.hostTriplet, 'tools')
+        cmakeTemplate = 'set(CMAKE_TOOLCHAIN_FILE "{}" CACHE FILEPATH "Toolchain file")\n'
+        cmakeTemplate += 'set(VKPKG_INSTALL_ROOT "{}" CACHE FILEPATH "vcpkg installed packages path")\n'
```
Comment (AndrewMeadows) [inline prebuild.py]: Why is it called `VKPKG_INSTALL_ROOT` instead of `VCPKG_INSTALLL_ROOT`?

File: cmake/ports/hifi-deps/usage [hunk 3]
```diff
@@ -0,0 +1 @@
+Hello!
```
Comment (AndrewMeadows) [inline cmake/ports/hifi-deps/usage]: Not a blocker: what is up with these `usage` files with the single line of `Hello!` ?

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/14927/HighFidelity-Beta-PR14297-556a857.apk

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14927/HighFidelity-Beta-Interface-PR14297-556a857.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14927/HighFidelity-Beta-PR14297-556a857.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14927/HighFidelity-Beta-Interface-PR14297-556a857.exe

File: prebuild.py (hunk 2 above)
Comment (jherico) [reply prebuild.py]: VCPKG_INSTALLL_ROOT has too many Ls?

File: cmake/ports/hifi-deps/usage (hunk 3 above)
Comment (jherico) [reply cmake/ports/hifi-deps/usage]: I patterned the creation of my 'meta pacakge' off the existing OpenSSL package.  They're probably not necessary.  I'll see if I can remove them.

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/14942/HighFidelity-Beta-PR14297-8c4f0ec.apk

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/14943/HighFidelity-Beta-PR14297-2691c39.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14943/HighFidelity-Beta-PR14297-2691c39.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14943/HighFidelity-Beta-Interface-PR14297-2691c39.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14943/HighFidelity-Beta-PR14297-2691c39.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14943/HighFidelity-Beta-Interface-PR14297-2691c39.exe

**Comment by qlimnavarro** [comment]
Currently Testing

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/14951/HighFidelity-Beta-PR14297-3d2e723.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14951/HighFidelity-Beta-PR14297-3d2e723.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14951/HighFidelity-Beta-Interface-PR14297-3d2e723.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/14960/HighFidelity-Beta-PR14297-5f6369d.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14960/HighFidelity-Beta-PR14297-5f6369d.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14960/HighFidelity-Beta-Interface-PR14297-5f6369d.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14960/HighFidelity-Beta-PR14297-5f6369d.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14960/HighFidelity-Beta-Interface-PR14297-5f6369d.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/14971/HighFidelity-Beta-PR14297-3e0dc35.apk

**Comment by qaAdamSmith82** [comment]
FAIL @conklin94122  This PR failed 58% of the test in NitPick. 

I have double checked my results against master, and the tests pass in master.

Test Results:
https://highfidelity.testrail.net/index.php?/runs/view/7510

Actual/Expected images 
https://drive.google.com/open?id=1IDA__KOf7fnVWgm3FMy81D2d3cPjdBs-

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14971/HighFidelity-Beta-PR14297-3e0dc35.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14971/HighFidelity-Beta-Interface-PR14297-3e0dc35.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14971/HighFidelity-Beta-PR14297-3e0dc35.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14971/HighFidelity-Beta-Interface-PR14297-3e0dc35.exe

**Comment by jherico** [comment]
seeing the same failures, but only when I run the automated tests.  When I run tests manually, they work.  Investigating.  This is a strange thing for the build system change to affect.

**Comment by qlimnavarro** [comment]
Currently Pre-Merge testing

**Comment by qlimnavarro** [comment]
Pre-Merge Pass @conklin94122 
Basic Rendering tests pass

**Pull request merged**